Middleware crashes websocket connections with `stdlib._net_http_ResponseWriter is not http.Hijacker: missing method Hijack`

Traefik plugins that wrap `ResponseWriter` without implementing `Hijack` and `Flush` crash websocket connections. Essentially the same issue as https://github.com/SchmitzDan/traefik-plugin-proxy-cookie/issues/3.

See also:

- https://github.com/traefik/traefik/issues/8682
- https://github.com/traefik/yaegi/issues/1425
- https://github.com/traefik/traefik/issues/8682 

## Steps to reproduce

1. `git clone https://github.com/safing/plausiblefeeder && cd plausiblefeeder`
2. Install [wscat](https://github.com/websockets/wscat): `npm i -g wscat`
3. `cd testdata && docker compose up`
4. In another terminal window, run `wscat -c ws://localhost:80/echo`
5. Middleware panics with: `"Recovered from panic in HTTP handler [172.19.0.1:47724 - /echo]: interface conversion: stdlib._net_http_ResponseWriter is not http.Hijacker: missing method Hijack"`

<details>
<summary>Full crash logs</summary>

```txt
testdata-traefik-1              | time="2023-07-20T18:17:35Z" level=info msg="Configuration loaded from flags."
testdata-traefik-1              | debug: converting report extension to ".html"
testdata-traefik-1              | creating plausiblefeeder plugin "my-plausiblefeeder@file" with config: &{EventEndpoint:http://plausible:8000/api/event Domains:[localhost] ReportExtensions:[.html ] ReportAllResources:false ReportAnyHost:false ReportErrors:false RemoteIPFromHeader:X-Forwarded-For QueueSize:1000 DebugLogging:true}
testdata-traefik-1              | time="2023-07-20T18:17:38Z" level=error msg="plugins-local/src/github.com/safing/plausiblefeeder/main.go:131:6: panic" plugin=plugin-plausiblefeeder module=github.com/safing/plausiblefeeder
testdata-traefik-1              | time="2023-07-20T18:17:38Z" level=error msg="Recovered from panic in HTTP handler [172.19.0.1:47724 - /echo]: interface conversion: stdlib._net_http_ResponseWriter is not http.Hijacker: missing method Hijack" middlewareName=traefik-internal-recovery middlewareType=Recovery
testdata-traefik-1              | time="2023-07-20T18:17:38Z" level=error msg="Stack: goroutine 89 [running]:\ngithub.com/traefik/traefik/v2/pkg/middlewares/recovery.recoverFunc({0x53d60e0, 0xc001aba000}, 0xc000c92d00)\n\tgithub.com/traefik/traefik/v2/pkg/middlewares/recovery/recovery.go:46 +0x225\npanic({0x3eb23a0, 0xc00062e330})\n\truntime/panic.go:884 +0x213\ngithub.com/traefik/yaegi/interp.runCfg.func1()\n\tgithub.com/traefik/yaegi@v0.15.0/interp/run.go:205 +0x1a5\npanic({0x3eb23a0, 0xc00062e330})\n\truntime/panic.go:884 +0x213\ngithub.com/traefik/traefik/v2/pkg/middlewares/pipelining.(*writerWithoutCloseNotify).Hijack(0xc000c93100?)\n\tgithub.com/traefik/traefik/v2/pkg/middlewares/pipelining/pipelining.go:69 +0x2c\nnet/http/httputil.(*ReverseProxy).handleUpgradeResponse(0xc000f5b590, {0x53c8cd0, 0xc0011be380}, 0xc000c93100, 0xc001488000)\n\tnet/http/httputil/reverseproxy.go:766 +0x584\nnet/http/httputil.(*ReverseProxy).ServeHTTP(0xc000f5b590, {0x53c8cd0?, 0xc0011be380}, 0xc000c92f00)\n\tnet/http/httputil/reverseproxy.go:484 +0x1051\ngithub.com/traefik/traefik/v2/pkg/middlewares/pipelining.(*pipelining).ServeHTTP(0xc0008fd6d0, {0x7f681c22aa80?, 0xc00053f5c0}, 0xc000c92f00)\n\tgithub.com/traefik/traefik/v2/pkg/middlewares/pipelining/pipelining.go:36 +0xc9\ngithub.com/traefik/traefik/v2/pkg/middlewares/accesslog.(*FieldHandler).ServeHTTP(0xc00149a680, {0x7f681c22aa80, 0xc00053f5c0}, 0xc001aec3a0?)\n\tgithub.com/traefik/traefik/v2/pkg/middlewares/accesslog/field_middleware.go:31 +0x122\ngithub.com/vulcand/oxy/v2/roundrobin.(*RoundRobin).ServeHTTP(0xc000f95d00, {0x7f681c22aa80, 0xc00053f5c0}, 0xc000c92e00)\n\tgithub.com/vulcand/oxy/v2@v2.0.0-20230227135449-a0e9f7ff1040/roundrobin/rr.go:102 +0x516\ngithub.com/traefik/traefik/v2/pkg/middlewares/emptybackendhandler.(*emptyBackend).ServeHTTP(0xc0008fd730, {0x7f681c22aa80, 0xc00053f5c0}, 0x44b84e0?)\n\tgithub.com/traefik/traefik/v2/pkg/middlewares/emptybackendhandler/empty_backend_handler.go:25 +0xec\ngithub.com/traefik/traefik/v2/pkg/middlewares/tracing.(*forwarderMiddleware).ServeHTTP(0xc00079ac30, {0x7f681c22aa80?, 0xc00053f5c0}, 0xc000c92e00)\n\tgithub.com/traefik/traefik/v2/pkg/middlewares/tracing/forwarder.go:38 +0x482\nreflect.Value.call({0x3eecf20?, 0xc00079b050?, 0x44b84e0?}, {0x48a08f8, 0x4}, {0xc00053f470, 0x2, 0x30?})\n\treflect/value.go:586 +0xb0b\nreflect.Value.Call({0x3eecf20?, 0xc00079b050?, 0x454432?}, {0xc00053f470?, 0x483eba0?, 0xc001aca420?})\n\treflect/value.go:370 +0xbc\ngithub.com/traefik/yaegi/interp.callBin.func2({0x3eecf20?, 0xc00079b050?, 0xc001aecc30?}, {0xc00053f470?, 0xc000115860?, 0xc00079b050?})\n\tgithub.com/traefik/yaegi@v0.15.0/interp/run.go:1506 +0x28\ngithub.com/traefik/yaegi/interp.callBin.func11(0xc001aca420)\n\tgithub.com/traefik/yaegi@v0.15.0/interp/run.go:1683 +0x15f\ngithub.com/traefik/yaegi/interp.runCfg(0xc000ee9cc0, 0xc001aca420, 0x6?, 0x4723920?)\n\tgithub.com/traefik/yaegi@v0.15.0/interp/run.go:213 +0x29d\ngithub.com/traefik/yaegi/interp.genFunctionWrapper.func1.1({0xc00053f140, 0x2, 0x4?})\n\tgithub.com/traefik/yaegi@v0.15.0/interp/run.go:1015 +0x825\ngithub.com/traefik/yaegi/stdlib._net_http_Handler.ServeHTTP(...)\n\tgithub.com/traefik/yaegi@v0.15.0/stdlib/go1_20_net_http.go:292\ngithub.com/traefik/traefik/v2/pkg/middlewares/accesslog.(*FieldHandler).ServeHTTP(0xc00149a940, {0x53d60e0, 0xc001aba000}, 0x3dc5460?)\n\tgithub.com/traefik/traefik/v2/pkg/middlewares/accesslog/field_middleware.go:31 +0x122\ngithub.com/gorilla/mux.(*Router).ServeHTTP(0xc0008faf00, {0x53d60e0, 0xc001aba000}, 0xc000c92d00)\n\tgithub.com/gorilla/mux@v1.8.0/mux.go:141 +0x24c\ngithub.com/traefik/traefik/v2/pkg/middlewares/recovery.(*recovery).ServeHTTP(0x30?, {0x53d60e0?, 0xc001aba000?}, 0xc000ce88d0?)\n\tgithub.com/traefik/traefik/v2/pkg/middlewares/recovery/recovery.go:32 +0x82\ngithub.com/traefik/traefik/v2/pkg/middlewares/accesslog.(*FieldHandler).ServeHTTP(0xc00149ab00, {0x53d60e0, 0xc001aba000}, 0x100?)\n\tgithub.com/traefik/traefik/v2/pkg/middlewares/accesslog/field_middleware.go:31 +0x122\ngithub.com/traefik/traefik/v2/pkg/middlewares.(*HTTPHandlerSwitcher).ServeHTTP(0x410d47?, {0x53d60e0, 0xc001aba000}, 0x5388401?)\n\tgithub.com/traefik/traefik/v2/pkg/middlewares/handler_switcher.go:23 +0x62\ngithub.com/traefik/traefik/v2/pkg/middlewares/requestdecorator.(*RequestDecorator).ServeHTTP(0xc00011afd8, {0x53d60e0, 0xc001aba000}, 0xc000c92800, 0xc0006337d8)\n\tgithub.com/traefik/traefik/v2/pkg/middlewares/requestdecorator/request_decorator.go:47 +0x30e\ngithub.com/traefik/traefik/v2/pkg/middlewares/requestdecorator.WrapHandler.func1.1({0x53d60e0?, 0xc001aba000?}, 0xc00053ef90?)\n\tgithub.com/traefik/traefik/v2/pkg/middlewares/requestdecorator/request_decorator.go:89 +0x68\nnet/http.HandlerFunc.ServeHTTP(0xc000f735e0?, {0x53d60e0?, 0xc001aba000?}, 0xc000f5a5a0?)\n\tnet/http/server.go:2122 +0x2f\ngithub.com/traefik/traefik/v2/pkg/middlewares/forwardedheaders.(*XForwarded).ServeHTTP(0xc000f735e0, {0x53d60e0, 0xc001aba000}, 0xc000c92800)\n\tgithub.com/traefik/traefik/v2/pkg/middlewares/forwardedheaders/forwarded_header.go:192 +0xca\nnet/http.AllowQuerySemicolons.func1({0x53d60e0, 0xc001aba000}, 0xc000c92800)\n\tnet/http/server.go:2963 +0x223\nnet/http.HandlerFunc.ServeHTTP(0x41e2b1?, {0x53d60e0?, 0xc001aba000?}, 0x410968?)\n\tnet/http/server.go:2122 +0x2f\ngolang.org/x/net/http2/h2c.h2cHandler.ServeHTTP({{0x53b3220?, 0xc000117950?}, 0xc000f73630?}, {0x53d60e0, 0xc001aba000}, 0xc000c92800)\n\tgolang.org/x/net@v0.7.0/http2/h2c/h2c.go:125 +0x59b\nnet/http.serverHandler.ServeHTTP({0xc00053ef30?}, {0x53d60e0, 0xc001aba000}, 0xc000c92800)\n\tnet/http/server.go:2936 +0x316\nnet/http.(*conn).serve(0xc00088a630, {0x53d7c30, 0xc00062e090})\n\tnet/http/server.go:1995 +0x612\ncreated by net/http.(*Server).Serve\n\tnet/http/server.go:3089 +0x5ed\n" middlewareType=Recovery middlewareName=traefik-internal-recovery

```

## Expected behavior

Websocket connections proxied through traefik should work with the plugin.

</details>

--- a/main.go
+++ b/main.go
@@ -133,7 +133,7 @@ func (pef *PlausibleEventFeeder) ServeHTTP(w http.ResponseWriter, r *http.Reques
 		return
 	}
 
-	// If the resource should be reported, we warp the response writer and check the status code before reporting.
+	// If the resource should be reported, we wrap the response writer and check the status code before reporting.
 	wrappedResponseWriter := &ResponseWriter{
 		ResponseWriter: w,
 		request:        r,

--- a/responsewriter.go
+++ b/responsewriter.go
@@ -1,6 +1,11 @@
 package plausiblefeeder
 
-import "net/http"
+import (
+	"bufio"
+	"fmt"
+	"net"
+	"net/http"
+)
 
 // ResponseWriter is used to wrap given response writers.
 type ResponseWriter struct {
@@ -18,4 +23,19 @@ func (rw *ResponseWriter) WriteHeader(code int) {
 
 	// Continue with the original method.
 	rw.ResponseWriter.WriteHeader(code)
+}
+
+func (rw *ResponseWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	hijacker, ok := rw.ResponseWriter.(http.Hijacker)
+	if !ok {
+		return nil, nil, fmt.Errorf("%T is not a http.Hijacker", rw.ResponseWriter)
+	}
+
+	return hijacker.Hijack()
+}
+
+func (rw *ResponseWriter) Flush() {
+	if flusher, ok := rw.ResponseWriter.(http.Flusher); ok {
+		flusher.Flush()
+	}
 }

